### ISSUE TITLE
TelegramApiErrorResponseException: null while GetUpdates Request

### ISSUE THREAD

**Issue by bratinghosh** [rubenlagus/TelegramBots#1401]
**Describe the bug**
I have been noticing TelegramApiErrorResponseException if I leave the long polling bot running for couple of days without any updates.

Surprisingly, if i restart the application, the error goes away. Any idea if this is a persistent issue?

**To Reproduce**
Starting the long polling bot application and leave it running for couple of days without activity.

**Expected behavior**
The bot is expected to receive 0 updates without any exception as no new updates are created during that period.

**Screenshots**
![Screenshot 2024-07-08 at 3 18 37 PM](https://github.com/rubenlagus/TelegramBots/assets/46055279/85dd3694-2e0a-482b-ae41-c105507ea590)
roblem.

**Comment by aNNiMON** [comment]
The problem on Telegram servers end, or with the network when reaching them.  Here's mine logs: 

```log
2024-06-14 04:11:24.694 [pool-3-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 647 millis...
2024-06-16 04:11:22.187 [pool-5-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 327 millis...
2024-06-17 02:28:18.716 [pool-7-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 655 millis...
2024-06-17 02:45:52.528 [pool-7-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 463 millis...
2024-06-25 04:10:43.180 [pool-3-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 726 millis...
2024-06-26 04:10:15.938 [pool-7-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 579 millis...
2024-07-01 04:46:14.304 [pool-7-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 694 millis...
2024-07-07 02:08:43.632 [pool-7-thread-1] ERROR org.telegram.telegrambots.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 451 millis...
```

**Comment by bratinghosh** [comment]
I do receive the occasional Exceptions and then resets to usual behaviour. But for the case posted above, it is quite peculiar as it reaches the backoff limit of 90000 millis and any getUpdate after always leads to an Exception. But if backoff limit is not reached, most of the time the retries reset to normal behaviour without exception. 

The only way I have managed to reproduce this issue is to keep the bot idle and not give it any updates for several days (1 week to be safe).

**Comment by aNNiMON** [comment]
@brating actually, I've never reached such timeouts. Sometimes Telegram servers are not available for 2-4 minutes, the maximum is 12830 according to my logs.
I also tried to reproduce with maximum timeout, but after that the updates are received as well. 

I guess it's something on your end. Possibly network configuration. Also try to update the library to the latest version and/or enable debug logs for a week and check them later.

**Comment by nemo83** [comment]
I'm getting a good few of these too.

I'm running on kubernetes on a home network. Ideas?

```2025-07-16T13:43:26.620Z ERROR 1 --- [pool-6-thread-1] o.t.telegrambots.longpolling.BotSession  : Error received from Telegram GetUpdates Request, retrying in 451 millis...
org.telegram.telegrambots.longpolling.exceptions.TelegramApiErrorResponseException: null
	at org.telegram.telegrambots.longpolling.BotSession.getUpdatesFromTelegram(BotSession.java:156) ~[telegrambots-longpolling-8.0.0.jar!/:na]
	at org.telegram.telegrambots.longpolling.BotSession.lambda$createPollerTask$1(BotSession.java:103) ~[telegrambots-longpolling-8.0.0.jar!/:na]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
Caused by: java.net.SocketTimeoutException: timeout
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.newTimeoutException(Http2Stream.kt:675) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http2.Http2Stream$StreamTimeout.exitAndThrowIfTimedOut(Http2Stream.kt:684) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http2.Http2Stream.takeHeaders(Http2Stream.kt:143) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http2.Http2ExchangeCodec.readResponseHeaders(Http2ExchangeCodec.kt:97) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.Exchange.readResponseHeaders(Exchange.kt:110) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.kt:93) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:34) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201) ~[okhttp-4.12.0.jar!/:na]
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154) ~[okhttp-4.12.0.jar!/:na]
	at org.telegram.telegrambots.longpolling.BotSession.getUpdatesFromTelegram(BotSession.java:139) ~[telegrambots-longpolling-8.0.0.jar!/:na]
	... 7 common frames omitted```

**Comment by webcane** [comment]
the issue is still where in v.9.0.0
sometimes it affects bot execution sometimes not.
```
2025-08-08 11:34:30 ERROR [pool-3-thread-1] o.t.t.longpolling.BotSession - Error received from Telegram GetUpdates Request, retrying in 624 millis...
org.telegram.telegrambots.longpolling.exceptions.TelegramApiErrorResponseException: null
	at org.telegram.telegrambots.longpolling.BotSession.getUpdatesFromTelegram(BotSession.java:156)
	at org.telegram.telegrambots.longpolling.BotSession.lambda$createPollerTask$1(BotSession.java:103)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
Caused by: org.telegram.telegrambots.longpolling.exceptions.TelegramApiErrorResponseException: 
	at org.telegram.telegrambots.longpolling.BotSession.getUpdatesFromTelegram(BotSession.java:152)
	... 7 common frames omitted

```

under the hood okHttpClient receive 409 status code.
```
Response{protocol=h2, code=409, message=, url=https://api.telegram.org/bot<token>/getupdates}
```

**Comment by EMIR1HUB** [comment]
I am using version 9.2.0, and this error is still occurring. 
Is the problem still in the library?

```
2025-10-23T22:50:40.108+05:00 ERROR [pool-3-thread-1] o.t.telegrambots.longpolling.BotSession  : Error received from Telegram GetUpdates Request, retrying in 514 millis...

org.telegram.telegrambots.longpolling.exceptions.TelegramApiErrorResponseException: null
	at org.telegram.telegrambots.longpolling.BotSession.getUpdatesFromTelegram(BotSession.java:156) ~[telegrambots-longpolling-9.2.0.jar:na]
	at org.telegram.telegrambots.longpolling.BotSession.lambda$createPollerTask$1(BotSession.java:103) ~[telegrambots-longpolling-9.2.0.jar:na]
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[na:na]
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java:358) ~[na:na]
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:305) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[na:na]
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[na:na]
	at java.base/java.lang.Thread.run(Thread.java:1583) ~[na:na]
Caused by: org.telegram.telegrambots.longpolling.exceptions.TelegramApiErrorResponseException: 
	at org.telegram.telegrambots.longpolling.BotSession.getUpdatesFromTelegram(BotSession.java:152) ~[telegrambots-longpolling-9.2.0.jar:na]```

**Comment by nemo83** [comment]
Hey everyone, 

just for context, I run the bot in docker container, on a single node k3s cluster.

When I was running it on a server at home (fiber optic, but behind residential-grade router) the disconnections would be frequent and long.

Same ubuntu version, same k3s version and same docker image on a cloud bare metal instance, and all the issues are gone. Like gone gone.

I'm 100% sure my "carrier branded" router does some crazy shit with DNS look up, in fact I had to disable dhcp for my home clusters, but although all DNS resolution were fixed and fast, I kept having long connections being randomly terminated. 

So tl;dr, moving workload to a well managed cloud server (still cheap, self managed, bare metal, non-vm instance) fixed the problem for me.

**Comment by valdisdot** [comment]
Got the same with 9.5.0, I believe the problem is OkHttpClient missuse in TelegramBotsLongPollingApplication constructor (there is no constructor options without http client if the only need is to set a custom Supplier of ScheduledExecutorService)

Have the problem when construct TelegramBotsLongPollingApplication like:
```java
            application = new TelegramBotsLongPollingApplication(
                    () -> objectMapper,
                    new OkHttpClient.Builder()::build,
                    scheduledExecutorServiceSupplier
            );
```

Have no problem while getting client from TelegramOkHttpClientFactory:
```java
            application = new TelegramBotsLongPollingApplication(
                    () -> objectMapper,
                    new TelegramOkHttpClientFactory.DefaultOkHttpClientCreator(),
                    scheduledExecutorServiceSupplier
            );
```

Or with a simply built client with proper timeouts:
```java
            application = new TelegramBotsLongPollingApplication(
                    () -> objectMapper,
                    new OkHttpClient.Builder()
                            .callTimeout(60, TimeUnit.SECONDS)
                            .connectTimeout(60, TimeUnit.SECONDS)
                            .readTimeout(60, TimeUnit.SECONDS)
                            .writeTimeout(60, TimeUnit.SECONDS)
                            .retryOnConnectionFailure(true)
                            .followRedirects(true)
                            ::build,
                    scheduledExecutorServiceSupplier
            );
```
